### PR TITLE
feat: Kube API Test

### PR DESCRIPTION
**Pull request by csviri**
## Description

Moving and renaming [Jenvtest](https://github.com/java-operator-sdk/jenvtest) here.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by csviri** [comment]
TODO for myself:
- [ ] Add documentation
- [ ] Update changelog

File: pom.xml
Comment (manusa) [inline pom.xml:126]: We'll eventually want to unify the logging system

File: pom.xml
Comment (manusa) [inline pom.xml:214]: should be added together with the Java11 modules, otherwise Java8 compilation won't work

File: junit/kube-api-test/client-inject/pom.xml
Comment (manusa) [inline junit/kube-api-test/client-inject/pom.xml:37]: Should not be needed.

Define the dependency in the dependencyManagement section of the main pom.xml

This will also help in generating the BOM

File: junit/kube-api-test/client-inject/src/main/java/io/fabric8/kubeapitest/junit/Fabric8ClientInjectionHandler.java
Comment (manusa) [inline junit/kube-api-test/client-inject/src/main/java/io/fabric8/kubeapitest/junit/Fabric8ClientInjectionHandler.java:50]: Note for self:
Once merged create issue to track the creation of a common junit module that provides the tooling to set fields, especially with the consideration for `@Nested`

File: junit/kube-api-test/core/pom.xml
Comment (manusa) [inline junit/kube-api-test/core/pom.xml:43]: Are these dependencies really required?
Commons compress is one of our optional dependencies, so I'm more specifically asking for commons-io.
In general, we try to reduce the number of dependencies. If this could be replaced with whatever we already have, it would be better.

File: junit/kube-api-test/core/pom.xml
Comment (manusa) [inline junit/kube-api-test/core/pom.xml:84]: version specification shouldn't be necessary

File: junit/kube-api-test/core/pom.xml
Comment (manusa) [inline junit/kube-api-test/core/pom.xml:96]: Note to self:
Curious about this one. Review current usage, see if can be easily replaced by Vert.x.

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java
Comment (manusa) [inline junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java:48]: Note that Vert.x provides this functionality, see `SelfSignedCertificate.create`

File: junit/kube-api-test/core/pom.xml
Comment (csviri) [reply junit/kube-api-test/core/pom.xml:96]: Can try to replace it with Very.x. Maybe a subsequent PR?

File: junit/kube-api-test/core/pom.xml
Comment (csviri) [reply junit/kube-api-test/core/pom.xml]: removed.

File: pom.xml
Comment (csviri) [reply pom.xml]: moved. thx

File: pom.xml
Comment (csviri) [reply pom.xml:126]: Note that this is just for tests

File: junit/kube-api-test/core/pom.xml
Comment (csviri) [reply junit/kube-api-test/core/pom.xml]: removed commons io, compress would be harder atm

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java
Comment (csviri) [reply junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java:48]: Are you ok to have a separate PR for that?

File: junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java
Comment (rohanKanojia) [inline junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java:37]: Instead of reusing static method for shared tests, do you think making an abstract class for simpleTest and making these classes extend that class would be better? Just like we do in httpclient tests...

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/TestUtils.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/TestUtils.java:29]: This class looks very much the same as `io.fabric8.kubeapitest.junit.TestUtils` . Is it possible to reduce duplication?

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/KubernetesMutationHookHandlingTest.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/KubernetesMutationHookHandlingTest.java:89]: Are we depending on Java Operator SDK's kubernetes-webhooks-framework for using AdmissionController class?  Could this lead to circular dependency issues? Shall we port the required class to Fabric8?

File: junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java
Comment (csviri) [reply junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java:37]: To be honest I don't really like that pattern, when we extending tests. But if everybody agrees it should be done that way in this project, I can change it.
I see it better with this "composition" over inheritance approach.

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/KubernetesMutationHookHandlingTest.java
Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/KubernetesMutationHookHandlingTest.java:89]: The core of Kube API Server does not use the client itself. I don't see how just porting that class would help.

File: pom.xml
Comment (manusa) [reply pom.xml:126]: Noted, but the idea is to keep dependencies to the minimum. Anyway, this is not urgent or critical.

File: junit/kube-api-test/core/pom.xml
Comment (manusa) [reply junit/kube-api-test/core/pom.xml]: Awesome, thx!

File: junit/kube-api-test/core/pom.xml
Comment (manusa) [reply junit/kube-api-test/core/pom.xml:90]: Yes, don't worry, this is just a reminder for me, especially with regards to https://github.com/fabric8io/kubernetes-client/pull/5632

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java
Comment (manusa) [reply junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/cert/CertManager.java:48]: Yes

File: junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java
Comment (manusa) [reply junit/kube-api-test/client-inject/src/test/java/io/fabric8/kubeapitest/junit/sample/JUnitFabric8ClientInjectPerMethodTest.java]: I don't like abstract tests either. However, I don't like the usage of this static test methods :).
Anyway, I'm sure we'll find a way to refactor the tests later on, especially if they are all doing the same.

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/Utils.java
Comment (manusa) [inline junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/Utils.java:83]: Might have issues with some JDK versions

See:
- https://github.com/eclipse/jkube/pull/2641
- https://github.com/eclipse/jkube/issues/2501

**Review by manusa — APPROVED** [review]
There's some logic duplication that we'll need to find a way yo unify in the future.
It might also be interesting to mark these modules as `(preview)` with a non-stable API, just in case we need to modify a few things once we start getting broader feedback.
Other than that, seems good to go for me.

File: doc/kube-api-test.md
Comment (rohanKanojia) [inline doc/kube-api-test.md:53]: I think these links need to be updated to be under `fabric8io` organization

File: doc/kube-api-test.md
Comment (rohanKanojia) [inline doc/kube-api-test.md:108]: It would be nice to have these available as a table (just like we have in [Configuring the Client](https://github.com/fabric8io/kubernetes-client#configuring-the-client) section)

File: junit/kube-api-test/pom.xml
Comment (rohanKanojia) [inline junit/kube-api-test/pom.xml:40]: Do we really need this module to be on JDK11 ? Most of the modules in this repository are based on JDK8. If not, we would also need to modify [release script](https://github.com/fabric8io/kubernetes-client/blob/main/.github/workflows/release.yaml#L100) to add this as list of java 11 modules to be released.

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java:45]:
```suggestion
                .contains(API_SERVER_PORT)
```

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/JUnitExtensionOnMethodTest.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/JUnitExtensionOnMethodTest.java:41]: Sorry, I haven't checked deeply but on a quick look I'm not able to figure out how `simpleTest1()` is testing something different than `simpleTest2()` . Or could it be a typo?

Comment (rohanKanojia) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/JUnitExtensionOnMethodTest.java:41]: If we want to deliberately repeat the execution, would[ `@RepeatedTest](https://junit.org/junit5/docs/5.4.1/api/org/junit/jupiter/api/RepeatedTest.html)` work in this case?

File: doc/kube-api-test.md
Comment (rohanKanojia) [inline doc/kube-api-test.md:1]: This document needs to be referenced in root README.md as well . Maybe we should add it below [`Mocking Kubernetes`](https://github.com/fabric8io/kubernetes-client?tab=readme-ov-file#mocking-kubernetes) section?

File: junit/kube-api-test/pom.xml
Comment (manusa) [reply junit/kube-api-test/pom.xml:40]: this was discussed at the internal meeting. It uses the JDK HTTP client.

Comment (rohanKanojia) [reply junit/kube-api-test/pom.xml:40]: @manusa : Do we need to modify release script to include this module in `Build and release Java 11 modules` step?

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/EnableKubeAPIServer.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/EnableKubeAPIServer.java:38]: Why not use string literal directly here?

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/KubeAPIServerExtension.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/KubeAPIServerExtension.java:87]:
```suggestion
                  "This is probably a misconfiguration since server won't be easily accessible.");
```

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/process/EtcdProcess.java
Comment (rohanKanojia) [inline junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/process/EtcdProcess.java:125]: This class seems to have some common logic as KubeAPIServerProcess, maybe it can be refactored to a common class.

File: doc/kube-api-test.md
Comment (csviri) [reply doc/kube-api-test.md:108]: I have mixed feelings about that, would help for sure, to my taste this is readable / usable enough but would worth it yes.

Comment (rohanKanojia) [reply doc/kube-api-test.md]: Umm, okay. It's totally up to you.

File: doc/kube-api-test.md
Comment (csviri) [reply doc/kube-api-test.md]: updated, thx!

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java
Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java:45]: This won't work, since server port is not a string

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/cert/CertManagerTest.java
Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/cert/CertManagerTest.java:30]: good point, that would be better, for sure

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java
Comment (rohanKanojia) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java:44]: Umm, why do we have static constants for this test when they're only used in one test?

Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java:44]: It's used in multiple places in the code (the same value), but also readablity is better (but this might be a personal thing)

Comment (rohanKanojia) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/kubeconfig/KubeConfigTest.java:44]: okay, I'm okay with this change.

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/cert/CertManagerTest.java
Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/cert/CertManagerTest.java]: fixed, thx

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/EnableKubeAPIServer.java
Comment (csviri) [reply junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/junit/EnableKubeAPIServer.java:37]: So it can be set using a constant when the annotation is used.

**Comment by csviri** [comment]
> There's some logic duplication that we'll need to find a way yo unify in the future. It might also be interesting to mark these modules as `(preview)` with a non-stable API, just in case we need to modify a few things once we start getting broader feedback. Other than that, seems good to go for me.

How should I mark this as `preview` exactly?

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/Utils.java
Comment (csviri) [reply junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/Utils.java]: ok, borrowed some code from jkube. should be fixed :)

File: doc/kube-api-test.md
Comment (csviri) [reply doc/kube-api-test.md:1]: added a section into the docs, pls check if this is ok this way. thx!

File: junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/process/EtcdProcess.java
Comment (csviri) [reply junit/kube-api-test/core/src/main/java/io/fabric8/kubeapitest/process/EtcdProcess.java:124]: yes, can I make a separate PR for this?

**Comment by csviri** [comment]
I think most of the points were addressed, the others will create PR in close future. Pls let me know if there is something else is blocking this PR (in case). thank you!

**Review by rohanKanojia — APPROVED** [review]

File: junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/JUnitExtensionOnMethodTest.java
Comment (csviri) [reply junit/kube-api-test/core/src/test/java/io/fabric8/kubeapitest/sample/JUnitExtensionOnMethodTest.java:40]: I think the intention was here to test if this whole stuff works with the repeated annotation, but it looks weird to me a bit, will revisit this too later if no objections.

**Comment by manusa** [comment]
Sorry :pray:, I forgot to merge and the PR got stale :(

Please, rebase and I'll merge as soon as the checks are green.

**Comment by csviri** [comment]
> Sorry 🙏, I forgot to merge and the PR got stale :(
> 
> Please, rebase and I'll merge as soon as the checks are green.

Hi rebased, the Openshift builds seem to be failing on all PR, not sure if the sonar results are good enough this way.

**Pull request merged**
